Description: Sharness: Set IFNAME prereq if network ifname found 
 Remove the "test -s ifname0.$$" statements from the sharness
 checks.  They make it difficult to diagnose why a check fails.
 If the file is empty, munged will fail with the error 'Failed to
 lookup origin ""' which provides useful information for debugging.
 Create the new check 'munged --origin interface name lookup' which
 greps the log from the preceding check for the loopback interface
 name, checks that it is not the empty string, saves the name to
 the file "ifname0.$$", and sets the sharness IFNAME prerequisite.
  Change the checks for 'munged --origin interface name' and
 'munged --origin interface name metadata' to depend on this new
 IFNAME prerequisite.  Thus, if munged is unable to match 127.0.0.1 to
 an interface name, these checks will be skipped instead of failing.
 Change _net_get_hostaddr_via_ifaddrs() to check that ifa_name
 is not the empty string before assigning the string for the network
 interface.
Author: Chris Dunlap <cdunlap@llnl.gov>
Origin: upstream, https://github.com/dun/munge/commit/77ff6823c423d19823d9259f8e0cae1fc98d9a7b
Last-Update: 2021-03-19
Forwarded: not-needed

--- a/src/munged/net.c
+++ b/src/munged/net.c
@@ -193,7 +193,9 @@ _net_get_hostaddr_via_ifaddrs (const char *name, struct in_addr *inaddrp,
      */
     if (ifa != NULL) {
         *inaddrp = ((struct sockaddr_in *) ifa->ifa_addr)->sin_addr;
-        *ifnamep = (ifa->ifa_name != NULL) ? strdup (ifa->ifa_name) : NULL;
+        *ifnamep = ((ifa->ifa_name != NULL) && (ifa->ifa_name[0] != '\0'))
+                ? strdup (ifa->ifa_name)
+                : NULL;
         rv = 0;
     }
     /*  If a match is not found, but host lookup succeeded...
diff --git a/t/0110-munged-origin-addr.t b/t/0110-munged-origin-addr.t
index 1e3f642..53bc5af 100755
--- a/t/0110-munged-origin-addr.t
+++ b/t/0110-munged-origin-addr.t
@@ -63,15 +63,12 @@ test_expect_success 'munged --origin null address warning' '
 '
 
 # Check if the origin address can be set by specifying an IP address.
-# Save the interface name to ifname0.$$ for later checks.
 ##
 test_expect_success 'munged --origin local IP address' '
     rm -f ifname0.$$ &&
     munged_start_daemon --origin=127.0.0.1 &&
     munged_stop_daemon &&
-    egrep "Set origin address to 127\.0\.0\.1\>" "${MUNGE_LOGFILE}" &&
-    sed -n -e "s/.*Set origin address.*(\([^)]*\)).*/\1/p" \
-            "${MUNGE_LOGFILE}" >ifname0.$$
+    egrep "Set origin address to 127\.0\.0\.1\>" "${MUNGE_LOGFILE}"
 '
 
 # Check if the origin address is set to the appropriate IP address in the
@@ -87,23 +84,35 @@ test_expect_success 'munged --origin local IP address metadata' '
     egrep "^ENCODE_HOST:.* 127\.0\.0\.1\>" meta.$$
 '
 
-# Check if the origin address can be set by specifying an interface name.
+# Check the log from the previous test for the network interface name
+#   corresponding to the loopback address.
+# Set the IFNAME prereq if "ifname0.$$" contains a non-empty string.
+##
+test_expect_success GETIFADDRS 'munged --origin interface name lookup' '
+    local ifname &&
+    sed -n -e "s/.*Set origin address.*(\([^)]*\)).*/\1/p" "${MUNGE_LOGFILE}" \
+            >ifname0.$$ &&
+    ifname=$(cat ifname0.$$) &&
+    test_debug "echo \"Loopback network interface name is [${ifname}]\"" &&
+    if test "x${ifname}" != x; then test_set_prereq IFNAME; fi
+'
+
+# Check if the origin address can be set by specifying the loopback network
+#   interface name.
 ##
-test_expect_success GETIFADDRS 'munged --origin interface name' '
-    test -s ifname0.$$ &&
+test_expect_success IFNAME 'munged --origin interface name' '
     munged_start_daemon --origin="$(cat ifname0.$$)" &&
     munged_stop_daemon &&
     egrep "Set origin address to 127\.0\.0\.1\>" "${MUNGE_LOGFILE}" &&
-    sed -n -e "s/.*Set origin address.*(\([^)]*\)).*/\1/p" \
-            "${MUNGE_LOGFILE}" >ifname1.$$ &&
+    sed -n -e "s/.*Set origin address.*(\([^)]*\)).*/\1/p" "${MUNGE_LOGFILE}" \
+            >ifname1.$$ &&
     test_cmp ifname0.$$ ifname1.$$
 '
 
 # Check if the origin address is set to the appropriate IP address in the
-#   credential metadata when specifying an interface name.
+#   credential metadata when specifying the loopback network interface name.
 ##
-test_expect_success GETIFADDRS 'munged --origin interface name metadata' '
-    test -s ifname0.$$ &&
+test_expect_success IFNAME 'munged --origin interface name metadata' '
     munged_start_daemon --origin="$(cat ifname0.$$)" &&
     "${MUNGE}" --socket="${MUNGE_SOCKET}" --no-input --output=cred.$$ &&
     "${UNMUNGE}" --socket="${MUNGE_SOCKET}" --input=cred.$$ \
-- 
2.31.0

